File: 参考/pyMarket/bot.py
import math
from random import random, uniform, choice
from typing import List
from decimal import Decimal, getcontext

import utils
from trader import Trader
from finance import Token, TradingPair

# 设置decimal精度
getcontext().prec = 28

chip_distribution = utils.ChipDistribution(0.05, 1, alpha=0.6)


class RandomBot(Trader):
    """随机交易机器人，继承自Trader类"""
    def __init__(self, name: str, trend: float, view: float):
        super().__init__(name)
        self.trading_pairs: List[TradingPair] = []
        self.view: Decimal = Decimal(str(view))  # 视图时间（使用Decimal）
        self.trend: Decimal = Decimal(str(trend))  # 趋势因子（使用Decimal）
        self.k = 0
    
    def set_trading_pairs(self, trading_pairs: List[TradingPair]) -> None:
        """设置机器人可以交易的交易对"""
        self.trading_pairs = trading_pairs
    
    def act(self):
        """执行随机交易动作"""
        if not self.trading_pairs:
            return False

        if len(self.orders) > 10 * len(self.trading_pairs):
            self.orders[0].close()

        
        # 先处理所有普通交易对，计算k的平均值
        for trading_pair in self.trading_pairs:
            action = random()
            if trading_pair.log:
                # 使用二分查找找到视图时间范围内的起始索引
                current_time = trading_pair.log[-1][0]
                target_time = current_time - float(self.view)

                # 二分查找找到第一个时间大于等于target_time的索引
                left, right = 0, len(trading_pair.log) - 1
                start_index = 0

                while left <= right:
                    mid = (left + right) // 2
                    if trading_pair.log[mid][0] >= target_time:
                        start_index = mid
                        right = mid - 1
                    else:
                        left = mid + 1

                # 确保start_index在有效范围内
                start_index = min(start_index, len(trading_pair.log) - 1)

                # 使用Decimal进行精确计算
                current_price = Decimal(str(trading_pair.log[-1][1]))
                start_price = Decimal(str(trading_pair.log[start_index][1]))
                price_change = (current_price - start_price) / start_price / self.view * self.trend
                new_k = utils.sigmoid((utils.sigmoid(float(price_change)) - 0.5)*10) - 0.5
                # print(new_k - self.k, self.k)
                sh = 0.5 + random() * 0.5
                action = action * 0.6 + float((new_k - self.k) * sh + new_k * (1 - sh))
                # print(action)
                ra = (random() + 3) / 4
                self.k = self.k * ra + new_k * (1 - ra)
            
            try:
                if action < 0.1:
                    self._place_market_order(trading_pair, 'sell')
                elif action < 0.5:
                    self._place_limit_order(trading_pair, 'sell')
                elif action < 0.9:
                    self._place_limit_order(trading_pair, 'buy')
                else:
                    self._place_market_order(trading_pair, 'buy')
            except Exception as e:
                print(f"机器人{self.name}普通交易失败: {e}")

    
    def _place_limit_order(self, trading_pair: TradingPair, direction: str) -> bool:
        """下单限价单"""
        if direction == 'buy':
            # 买入：基于持有计价代币的1%-20%
            d = random()
            ori_price = float(trading_pair.price)
            price = ori_price * (1 + d * math.e ** (-d / 0.05))
            quote_amount = float(self.assets.get(trading_pair.quote_token, Decimal('0')))
            if quote_amount <= 0:
                return False
            max_volume = quote_amount * chip_distribution.pdf(abs((price - ori_price) / ori_price)) * (random() + 3) / 4 / price

        else:  # 'sell'
            # 卖出：基于持有基准代币的1%-20%
            d = random()
            ori_price = float(trading_pair.price)
            price = ori_price * (1 - d * math.e ** (-d / 0.05))
            base_amount = float(self.assets.get(trading_pair.base_token, Decimal('0')))
            if base_amount <= 0:
                return False
            max_volume = base_amount * chip_distribution.pdf(abs((price - ori_price) / ori_price)) * (random() + 3) / 4
        
        volume = max(0.0001, max_volume)  # 确保交易量不为0
        # 提交订单
        order = self.submit(trading_pair, direction, price, volume)
        return order is not None
    
    def _place_market_order(self, trading_pair: TradingPair, direction: str) -> bool:
        """下单市价单"""
        # 根据持有资产计算随机交易量（市价单交易量较小）
        if direction == 'buy':
            # 买入：基于持有计价代币的0.5%-5%
            quote_amount = float(self.assets.get(trading_pair.quote_token, Decimal('0')))
            if quote_amount <= 0:
                return False

            max_volume = quote_amount * (1 - random() ** 3.5) / float(trading_pair.price)
        else:  # 'sell'
            # 卖出：基于持有基准代币的0.5%-5%
            base_amount = float(self.assets.get(trading_pair.base_token, Decimal('0')))
            if base_amount <= 0:
                return False
            max_volume = base_amount * (1 - random() ** 3.5)
        
        volume = max(0.000001, max_volume)  # 确保交易量不为0
        
        # 提交市价单
        # print((1 if direction == 'buy' else -1) * volume)
        return self.submit_market(trading_pair, direction, volume)


class BotManager:
    """机器人管理器，用于管理多个交易机器人"""
    def __init__(self):
        self.bots: List[RandomBot] = []
    
    def __len__(self) -> int:
        """返回机器人数量"""
        return len(self.bots)
    
    def create_bots_batch(self, count: int, asset_configs: dict, 
                         name_prefix: str = "Bot", trend: float = 0, view: float = 10) -> List[RandomBot]:
        """
        批量创建机器人并分配多种资产
        
        Args:
            count: 要创建的机器人数量
            asset_configs: 资产配置字典，格式为 {Token: {"min": min_amount, "max": max_amount}}
            name_prefix: 机器人名称前缀
        """
        for i in range(count):
            # 创建机器人实例
            bot = RandomBot(f"{name_prefix}_{i+1:03d}",
                            trend=trend*(-0.7+random()*1.5),
                            view=view*(0.1+random()*1.9))
            
            # 为每个代币分配随机资产
            for token, config in asset_configs.items():
                min_amount = config["min"]
                max_amount = config["max"]
                
                # 在最小值和最大值之间随机分配资产（将Decimal转换为float）
                amount = uniform(float(min_amount), float(max_amount))
                bot.add_asset(token, amount)
            
            self.bots.append(bot)

        return self.bots
    
    def create_bots(self, count: int, base_token: Token, quote_token: Token,
                    initial_base_amount: float = 1.0,
                    initial_quote_amount: float = 1000.0) -> List[RandomBot]:
        """创建指定数量的机器人（兼容旧版本）"""
        # 转换为新的资产配置格式
        asset_configs = {
            base_token: {"min": initial_base_amount * 0.9, "max": initial_base_amount * 1.1},
            quote_token: {"min": initial_quote_amount * 0.9, "max": initial_quote_amount * 1.1}
        }
        
        return self.create_bots_batch(count, asset_configs, "Bot")
    
    def set_trading_pairs(self, trading_pairs: List[TradingPair]) -> None:
        """为所有机器人设置可交易的交易对"""
        for bot in self.bots:
            bot.set_trading_pairs(trading_pairs)
    
    def step(self, trading_pairs: List[TradingPair]) -> None:
        """让所有机器人执行一步交易动作"""
        # 先设置交易对
        self.set_trading_pairs(trading_pairs)
        
        # 每个机器人执行交易动作
        for bot in self.bots:
            # 随机决定是否执行交易（70%概率）
            if random() < 0.7:
                bot.act()
    
    def get_average_asset_value(self, token: Token) -> float:
        """获取所有机器人某代币的平均持有量"""
        if not self.bots:
            return 0.0
        
        total = sum(bot.get_asset_value(token) for bot in self.bots)
        return total / len(self.bots)

File: rust_backend/src/bot.rs
use crate::finance::{Direction, TradingPair};
use crate::trader::Trader;
use crate::utils::{decimal_to_f64, sigmoid, ChipDistribution, f64_to_decimal};
use parking_lot::RwLock;
use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};
use rust_decimal::Decimal;
use std::collections::HashMap;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;

static BOT_ID_COUNTER: AtomicU64 = AtomicU64::new(1);

fn next_bot_id() -> u64 {
    BOT_ID_COUNTER.fetch_add(1, Ordering::SeqCst)
}

/// RandomBot implements random trading strategy
pub struct RandomBot {
    pub id: u64,
    pub trader: Arc<Trader>,
    pub trend: f64,
    pub view: f64,
    pub k: RwLock<f64>,
    pub trading_pairs: RwLock<Vec<Arc<TradingPair>>>,
    pub chip_dist: ChipDistribution,
    pub rng: RwLock<StdRng>,
}

impl RandomBot {
    pub fn new(name: impl Into<String>, trend: f64, view: f64) -> Arc<Self> {
        let trader = Trader::new(name);
        let seed = next_bot_id();

        Arc::new(Self {
            id: seed,
            trader,
            trend,
            view,
            k: RwLock::new(0.0),
            trading_pairs: RwLock::new(Vec::new()),
            chip_dist: ChipDistribution::new(0.05, 1.0, Some(0.6)),
            rng: RwLock::new(StdRng::seed_from_u64(seed)),
        })
    }

    pub fn set_trading_pairs(&self, pairs: Vec<Arc<TradingPair>>) {
        *self.trading_pairs.write() = pairs;
    }

    pub fn add_asset(&self, token_name: impl Into<String>, amount: Decimal) {
        self.trader.add_asset(token_name, amount);
    }

    pub fn act(&self) {
        let pairs = self.trading_pairs.read().clone();

        if pairs.is_empty() {
            return;
        }

        // Limit number of orders
        {
            let orders = self.trader.orders.read();
            let max_orders = 10 * pairs.len();
            if orders.len() > max_orders {
                // Cancel oldest order (simplified)
                // In real implementation, we'd need to track which trading pair
            }
        }

        for trading_pair in &pairs {
            let mut rng = self.rng.write();
            let action: f64 = rng.gen();
            drop(rng);

            let trade_log = trading_pair.trade_log.read();

            let mut final_action = action;

            if !trade_log.is_empty() {
                let current_time = trade_log.last().unwrap().timestamp;
                let target_time = current_time as f64 - self.view;

                // Find start index using binary search
                let start_index = Self::find_start_index(&trade_log, target_time as u64);

                let current_price = decimal_to_f64(trade_log.last().unwrap().price);
                let start_price = decimal_to_f64(trade_log[start_index].price);

                let price_change =
                    (current_price - start_price) / start_price / self.view * self.trend;
                let new_k = sigmoid((sigmoid(price_change) - 0.5) * 10.0) - 0.5;

                let mut k = self.k.write();
                let sh = 0.5 + self.rng.write().gen::<f64>() * 0.5;
                final_action = action * 0.6 + ((*k - new_k) * sh + new_k * (1.0 - sh));

                let ra = (self.rng.write().gen::<f64>() + 3.0) / 4.0;
                *k = *k * ra + new_k * (1.0 - ra);
                drop(k);
            }

            drop(trade_log);

            // Execute action
            let result = if final_action < 0.1 {
                self.place_market_order(trading_pair, Direction::Sell)
            } else if final_action < 0.5 {
                self.place_limit_order(trading_pair, Direction::Sell)
            } else if final_action < 0.9 {
                self.place_limit_order(trading_pair, Direction::Buy)
            } else {
                self.place_market_order(trading_pair, Direction::Buy)
            };

            if let Err(e) = result {
                // Silently fail for now
            }
        }
    }

    fn find_start_index(trade_log: &[crate::finance::TradeEntry], target_time: u64) -> usize {
        let mut left = 0;
        let mut right = trade_log.len() - 1;
        let mut start_index = 0;

        while left <= right {
            let mid = (left + right) / 2;
            if trade_log[mid].timestamp >= target_time {
                start_index = mid;
                if mid == 0 {
                    break;
                }
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        start_index.min(trade_log.len() - 1)
    }

    fn place_limit_order(
        &self,
        trading_pair: &Arc<TradingPair>,
        direction: Direction,
    ) -> Result<(), String> {
        let mut rng = self.rng.write();
        let d: f64 = rng.gen();
        drop(rng);

        let current_price = *trading_pair.price.read();
        let current_price_f64 = decimal_to_f64(current_price);

        let price = match direction {
            Direction::Buy => {
                let price_f64 = current_price_f64 * (1.0 + d * (-d / 0.05).exp());
                f64_to_decimal(price_f64)
            }
            Direction::Sell => {
                let price_f64 = current_price_f64 * (1.0 - d * (-d / 0.05).exp());
                f64_to_decimal(price_f64)
            }
        };

        let max_volume = match direction {
            Direction::Buy => {
                let quote_amount =
                    self.trader
                        .get_asset(&trading_pair.quote_token.name);
                if quote_amount <= Decimal::ZERO {
                    return Err("Insufficient quote token".to_string());
                }
                let price_diff = (price - current_price).abs() / current_price;
                let price_diff_f64 = decimal_to_f64(price_diff);
                let pdf = self.chip_dist.pdf(price_diff_f64);
                let mut rng = self.rng.write();
                let factor = (rng.gen::<f64>() + 3.0) / 4.0;
                drop(rng);

                quote_amount * f64_to_decimal(pdf * factor) / price
            }
            Direction::Sell => {
                let base_amount =
                    self.trader
                        .get_asset(&trading_pair.base_token.name);
                if base_amount <= Decimal::ZERO {
                    return Err("Insufficient base token".to_string());
                }
                let price_diff = (price - current_price).abs() / current_price;
                let price_diff_f64 = decimal_to_f64(price_diff);
                let pdf = self.chip_dist.pdf(price_diff_f64);
                let mut rng = self.rng.write();
                let factor = (rng.gen::<f64>() + 3.0) / 4.0;
                drop(rng);

                base_amount * f64_to_decimal(pdf * factor)
            }
        };

        let volume = max_volume.max(f64_to_decimal(0.0001));

        self.trader
            .submit(trading_pair, direction, price, volume)
            .ok_or("Failed to submit order".to_string())?;

        Ok(())
    }

    fn place_market_order(
        &self,
        trading_pair: &Arc<TradingPair>,
        direction: Direction,
    ) -> Result<(), String> {
        let mut rng = self.rng.write();
        let random_factor: f64 = rng.gen();
        drop(rng);

        let volume = match direction {
            Direction::Buy => {
                let quote_amount =
                    self.trader
                        .get_asset(&trading_pair.quote_token.name);
                if quote_amount <= Decimal::ZERO {
                    return Err("Insufficient quote token".to_string());
                }
                let current_price = *trading_pair.price.read();
                let volume_f64 =
                    decimal_to_f64(quote_amount) * (1.0 - random_factor.powf(3.5))
                        / decimal_to_f64(current_price);
                f64_to_decimal(volume_f64.max(0.000001))
            }
            Direction::Sell => {
                let base_amount =
                    self.trader
                        .get_asset(&trading_pair.base_token.name);
                if base_amount <= Decimal::ZERO {
                    return Err("Insufficient base token".to_string());
                }
                let volume_f64 =
                    decimal_to_f64(base_amount) * (1.0 - random_factor.powf(3.5));
                f64_to_decimal(volume_f64.max(0.000001))
            }
        };

        self.trader
            .submit_market(trading_pair, direction, volume)
            .ok_or("Failed to submit market order".to_string())?;

        Ok(())
    }

    pub fn get_asset_value(&self, token_name: &str) -> Decimal {
        self.trader.get_asset(token_name)
    }
}

/// BotManager manages multiple bots
pub struct BotManager {
    pub bots: RwLock<Vec<Arc<RandomBot>>>,
}

impl BotManager {
    pub fn new() -> Arc<Self> {
        Arc::new(Self {
            bots: RwLock::new(Vec::new()),
        })
    }

    pub fn create_bots_batch(
        &self,
        count: usize,
        asset_configs: &HashMap<String, (Decimal, Decimal)>,
        name_prefix: &str,
        trend: f64,
        view: f64,
    ) -> Vec<Arc<RandomBot>> {
        let mut bots = Vec::with_capacity(count);
        let mut rng = StdRng::from_entropy();

        for i in 0..count {
            let bot_trend = trend * (-0.7 + rng.gen::<f64>() * 1.5);
            let bot_view = view * (0.1 + rng.gen::<f64>() * 1.9);

            let bot = RandomBot::new(format!("{}_{:03}", name_prefix, i + 1), bot_trend, bot_view);

            // Allocate assets
            for (token_name, (min_amount, max_amount)) in asset_configs {
                let range = *max_amount - *min_amount;
                let amount = *min_amount + f64_to_decimal(rng.gen::<f64>()) * range;
                bot.add_asset(token_name.clone(), amount);
            }

            bots.push(bot);
        }

        let bot_refs: Vec<Arc<RandomBot>> = bots.iter().map(|b| Arc::clone(b)).collect();
        *self.bots.write() = bots;
        bot_refs
    }

    pub fn set_trading_pairs(&self, pairs: Vec<Arc<TradingPair>>) {
        for bot in self.bots.read().iter() {
            bot.set_trading_pairs(pairs.clone());
        }
    }

    pub fn step(&self) {
        for bot in self.bots.read().iter() {
            let mut rng = StdRng::from_entropy();
            if rng.gen::<f64>() < 0.7 {
                bot.act();
            }
        }
    }

    pub fn len(&self) -> usize {
        self.bots.read().len()
    }
}
